### ISSUE TITLE
Decouple PageObjects from Behat Implementation?

### ISSUE THREAD

**Issue by slde-gorillaman** [sensiolabs/BehatPageObjectExtension#61]
Hi there,

I have been thinking of using the page object pattern in a way to share the Objects with Symfony Webtestcases and found that the entire SensioLabs\Behat\PageObjectExtension\PageObject namespace could become an own library, making it possible to build Pageobjects for an application reusable among different test suites.

What do you think?

**Comment by jakzal** [comment]
I actually have done this partly few months ago. I just haven't finished nor published it.

I'm pretty busy till SymfonyCon. I'll be catching up with my open source stuff in December :)

**Comment by slde-gorillaman** [comment]
Awesome, looking forward to it :)

**Comment by jakzal** [comment]
I'm not sure if you'll be able to use it Symfony test cases easily yet as currently page objects are heavily based on Mink. It might be the next step though. I'll give it a thought.

**Comment by slde-gorillaman** [comment]
Mink has a BrowserKit Extension, this way it could be easily integrated

**Comment by slde-gorillaman** [comment]
As I see it, the building blocks are there, we only need to put them together and extract the PageObjects from this repo :)

**Comment by slde-gorillaman** [comment]
Here is an approach to integrate the Pop into Symfony WebTestCase. This is by no means a final approach, just a demo.

https://github.com/mozzymoz/e2eTests/tree/master/src/E2eTests/PopBundle

**Comment by DonCallisto** [comment]
Yes, that's interesting even for "non-test" scenarios: I would like to scrape a site for a service on which there's any rest API or WS in general.
I can use `Mink`, `Symfony` and this extension to obtain what I need but, sadly, this extension is too bounded to `Behat` and this would not be suitable for a Symfony Bundle.

@jakzal any thoughts?

**Comment by jakzal** [comment]
@DonCallisto I did some work which I haven't released in the end. I tried to make this in a real BC way, but since breaking changes in Mink (see #92), perhaps it's time to take this opportunity to rethink the design and come up with something new.

Page objects currently extend Mink elements, which was a cautious decision. The extensions initial focus group was testers with only basic coding experience. I don't think this is our user base at the moment.

One thing that's been in my head for quite some time is removing necessity to extend Mink elements, but provide tools to inject them directly into page object properties. We'd describe how to query elements with annotations:

```php
class SearchPage
{
    /**
     * @FindBy(name="search_box")
     */
    private $searchInput;

    /**
     * @FindBy(id="search")
     */
    private $submit;

    /**
     * @FindBy(css=".container .message")
     */
    private $message;

    /**
     * @FindBy(xpath="//div//h1")
     */
    private $title;
}
```

We could start fresh by implementing this approach as a library, and then integrate it with Behat in this extension in a next major version.

**Comment by DonCallisto** [comment]
Cool: only thing that "worry" me is about "collections", so dynamic selectors or something like this. 
Of course we can put placeholders and so on, of course, but I suppose it could require a studying phase before try this.

If you want, I'm in for doing some work with you.

Just let me know when and how 😄
